### ISSUE TITLE
sls deploy --stage production gives log group already exists

### ISSUE THREAD

**Issue by jatinvmehta** [serverless/serverless#5057]
sls deploy works fine and create cloud formation. When i run sls deploy --stage production, it gives error saying log group already exists. I don't configure custom log gorup

sls deploy --stage production
Serverless: Packaging service...
Serverless: Excluding development dependencies...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service .zip file to S3 (7.96 MB)...
Serverless: Validating template...
Serverless: Creating Stack...
Serverless: Checking Stack create progress...
.......
Serverless: Operation failed!

  Serverless Error ---------------------------------------

  An error occurred: MozendaDailyAgentLogGroup - /aws/lambda/mozenda-insertdailyagentscrappingdata-dev already exists in stack arn:aws:cloudformation:us-east-1:822605674378:stack/MozendaDataCollection-dev/36208d40-7324-11e8-8ced-503acac5c0fd.

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.11.3
     Serverless Version:     1.27.3

**Comment by sypannapis** [comment]
I also encounter this case when I tried to upgrade from Serverless v1.3 to v1.27.3, then I saw https://github.com/serverless/serverless/issues/3168. 
After I removed the log groups manually, the stack could be deployed successfully. Hope this helps.

**Comment by ndeitch** [comment]
Hey @jatinvmehta,

I was facing a similar problem, then reading [docs](https://serverless.com/framework/docs/providers/aws/guide/deploying#how-it-works-2) I realize that passing stage on deploy doesn't replace your package stage.

I hope it helps you.

**Comment by J-Rojas** [comment]
@ndeitch I'm facing this same problem and I don't understand what your workaround is (if any). The deploy command should repackage the files for the new stage, but are you suggesting there's a bug here?

**Comment by wongsean** [comment]
Same issue here

**Comment by kdavidmoore** [comment]
I get this message when deploying to any stage. We are redeploying some functions that used to exist in an old stack and kept the same function names. We deleted the old functions and their log groups using the AWS CLI. But serverless thinks the log group that I deleted is still there. When I try to delete the log group in question again using the AWS CLI, I get a ResourceNotFoundException, suggesting that the log group does not exist because I already deleted it. When I list the log groups using the AWS CLI, it is not there either.

**Comment by frangeris** [comment]
Guys this is a kind of blocker / stopper when having multiple environments, does anybody found a workaround for this?

@pmuens Do you know anything about this? 🙏 

Thanks you!

**Comment by J-Rojas** [comment]
If you are explicitly naming your Lambdas like this:

`name: myService-fooBar`

Then you should change that to:

`name: myService-${opt:stage}-fooBar`

This will fix any conflicting log group definitions. That's what solved my problem with this error. Hope that helps.

**Comment by occasl** [comment]
I just went to the CloudWatch Logs and deleted the log groups it said were already created. I redeployed and it was fine.

**Comment by pmuens** [comment]
Closing this for now since this issue is quite stale and we've shipped some fixes in the interim.

Feel free to continue the discussion here...

**Comment by tommy20061222** [comment]
i still have same issue, even after i delete the related log goups.

**Comment by frangeris** [comment]
@tommy20061222 use the _J-Rojas_, is the only way and the best for now, I've being using this way for various months and it really works


> If you are explicitly naming your Lambdas like this:
> 
> `name: myService-fooBar`
> 
> Then you should change that to:
> 
> `name: myService-${opt:stage}-fooBar`
> 
> This will fix any conflicting log group definitions between environments. That's what solved my problem with this error. Hope that helps.

**Comment by tommy20061222** [comment]
@frangeris I did it , here is the code:
 get:
    runtime: nodejs8.10
    name: 'helloworld-${self:provider.stage}'

**Comment by frangeris** [comment]
You need to have that value on `serverless.yml` like this:

```yml
service:
  name: myapi
  publish: false
provider:
  name: aws
  region: us-east-1
  runtime: nodejs8.10
  stage: develop
```

Or if you want to receive it via parameters (the coolest way) and `custom` :

```yml
custom:
  stage: '${opt:stage, self:provider.stage}'
```
Where this means:

_Use **stage** from options first, if not exists, then use the default value from **provider.stage**_

And via commands just pass it this way:

```bash
./node_modules/.bin/serverless deploy --stage develop|prod|staging
```

This will create a separate api for each environment you need, and not conflicts :)

Hope that helps 🍻

**Comment by accraze** [comment]
I was able to work around this by explicitly packaging and then deploying for each stage like this:

```
serverless package --stage prod
serverless deploy --package .serverless --stage prod
```

**Comment by frangeris** [comment]
@accraze does that create two different stage in the same api or two apis each one with its stage?

**Comment by michalorman** [comment]
> Or if you want to receive it via parameters (the coolest way) and `custom` :
> 
> ```yaml
> custom:
>   stage: '${opt:stage, self:provider.stage}'
> ```

Actually better configuration is:

```yaml
provider:
  stage: ${opt:stage, 'develop'}
```

**Comment by neotheicebird** [comment]
I think, this error has to do with the function names in `serverless.yaml`. Say for example, the `function` section is as follows:

```
functions:
  login:
    handler: handler.login
    name: login
    ...
```

This might cause a problem as multiple applications could have the same function name. Try making function names unique. I would do

```
functions:
  login:
    handler: handler.login
    name: ${self:provider.service}-${self:provider.stage}-login
    ...
```

By adding `service` name and `stage` as prefixes, it makes it highly unlikely to clash with functions from other applications. Hope this helps people who are still facing this issue.

**Comment by siolfyr** [comment]
I don't understand this issue at all. Deleting the log groups did nothing, the problem still persisted. "Log group already exists in stack" ... but no, it doesn't. Also, the stack printed out is my current stack, so why is that a problem? If the log groups already exist, then you're done, correct? Why try to recreate them?

To fix this, I just renamed the function to anything. `name: ${self:provider.stage}-${self:service}-function-name` became `name: ${self:provider.stage}-${self:service}-function-name2` ... then it worked. After deploying, changing it back to what it was originally worked fine.

Some sort of caching? I wonder what would have happened if I'd have deleted the .serverless directory instead of renaming the function like I did.

**Comment by joaomirandasa** [comment]
This issue still happening, LogGroupCreation does not include STAGE as part of his name so if you simply tried to change stage and deploy again you will face something like this:

`  Serverless Error ---------------------------------------
 
  An error occurred: NotificationReadLogGroup - /aws/lambda/*******-notificationRead already exists in stack arn:aws:cloudformation:sa-east-1:***************:stack/*******-dev/*******************************.`
  
  To avoid this behavior, mantainers must to include stage as part of **logGroupName**

**Comment by elleryfamilia** [comment]
Just ran into this issue out of nowhere when deploying to --stage dev after having done so successfully countless times. Frustrating that this issue was for reported over 2.5 years ago yet here we are.

**Comment by TomSinclair** [comment]
Is the current "fix" still to add the stage name to each individual function?

**Comment by shawnmclean** [comment]
Why do I need to package for each stage? CD should be: build once, deploy anywhere.

**Comment by pgrzesik** [comment]
@shawnmclean the issue is that during `package`, the underlying CloudFormation template is compiled, which depends on specific stage for naming, etc. Contents of produced zip artifacts can also change based on stage-dependent configuration.

**Comment by shawnmclean** [comment]
>Contents of produced zip artifacts can also change based on stage-dependent configuration.
Do you mean on `sls deploy` I'm able to pass the stage and variables and that will overwrite the stage I packaged with? 

This I couldn't get to work. My package functions still had the name with `dev` in them.

**Comment by pgrzesik** [comment]
@shawnmclean Sorry for not being 100% clear. What I meant was that since `stage` influences the output of `package` step, it cannot really be reused between stages. While the function ZIP artifacts potentially could be reused, the CloudFormation template cannot, that's why the package step has to be run for each stage anyway.

**Comment by Antontelesh** [comment]
Faced the same issue today, tried to manually remove log group and it was still failing.
Then I decided to pay closer attention and found that after deleting a log group it appears again in a few minutes.

My assumption is that during deployment the service is being used by someone and whenever there's an event triggered AWS tries to log it to CloudWatch, finds that a log group does not exist and creates it automatically.

I had to remove the log group and watch for it to not appear during the deployment. When it appeared once or twice I quickly removed it again and then my deployment succeeded.

I believe, this information may be interesting to someone
